### PR TITLE
Adds createEventStream and support for event stream responses

### PR DESCRIPTION
**Pull request by chrisradek**
Adds event stream unmarshalling.

In node.js version 0.10 and higher, an output shape modeled as an `eventstream` will be exposed as a readable stream that emits individual, parsed events on each `data` call. In node.js v10.x and higher since streams have experimental async iterator support they can be read in an async function while still maintaining back-pressure.

In browser environments, and nodejs 0.8, output shapes modeled as an `eventstream` is exposed as an array of events. In these cases the response is fully buffered before the events can be read.

I didn't include the S3 models with S3 select in this PR (will have to include them in the TS or Doc PR) but can if you want to test these changes with a real API.

For someone writing an isomorphic tool, they could simply check if the `Payload` object is an array, and assume it is a stream otherwise.

Some examples of how you'd use the S3 Select API:

```javascript
// node 10+ (also valid on regular arrays, so can be used in browsers)
async function test() {
  const result = await s3.selectObjectContent({/** params **/}).promise();
  const events = result.Payload;

  for await (const event of events) {
    if (event.Records) {
      console.log(event.Records.Payload.toString());
    } else {
      console.log(event);
    }
  }
  console.log('Done processing events');
}
```

```javascript
// node 0.10.x +
function test() {
  s3.selectObjectContent({/** params **/}, function (err, data) {
    if (err) {
      return console.error(err);
    }
    var eventStream = data.Payload;
    eventStream.on('data', function(event) {
      if (event.Records) {
        console.log(event.Records.Payload.toString());
      } else {
        console.log(event);
      }
    });
    eventStream.on('end', function() {
      console.log('Done processing events');
    });
  });
}
```

```javascript
// browsers or node 0.8
function test() {
  s3.selectObjectContent({/** params **/}, function (err, data) {
    if (err) {
      return console.error(err);
    }
    var events = data.Payload;
    for (var event of events) {
      if (event.Records) {
        console.log(event.Records.Payload.toString());
      } else {
        console.log(event);
      }
    }
  });
}
```

### CODE REVIEW

File: lib/event-stream/event-message-unmarshaller-stream.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+var Transform = require('stream').Transform;
+var parseEvent = require('./parse-event').parseEvent;
+
+/** @type {Transform} */
+function EventUnmarshallerStream(options) {
+    options = options || {};
+    // set output to object mode
+    options.readableObjectMode = true;
+    Transform.call(this, options);
+    this._readableState.objectMode = true;
```
Comment (AllanFly120) [inline lib/event-stream/event-message-unmarshaller-stream.js]: Is this duplicate with `options.readableObjectMode = true;`?

File: lib/event-stream/format-message.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
+var crypto = require('../../lib/util').crypto;
+var Int64 = require('./int64').Int64;
+var toBuffer = require('./to-buffer').toBuffer;
+var allocBuffer = require('./alloc-buffer').allocBuffer;
+
+function formatMessage(message) {
```
Comment (AllanFly120) [inline lib/event-stream/format-message.js]: This name is a bit confusing to me. `buildMessage` is probably a better name.

Comment (chrisradek) [reply lib/event-stream/format-message.js]: Sure, I don't have a strong preference either way.

File: lib/event_listeners.js [hunk 3]
```diff
@@ -264,7 +277,11 @@ AWS.EventListeners = {
         });
 
         httpResp.on('end', function onEnd() {
-          if (!stream || !stream.didCallback) {
+          if ((AWS.HttpClient.streamsApiVersion !== 2) ||
+            // don't concat response chunks when using event streams with streams unless response is unsuccessful
+            (!(operation.hasEventOutput && service.successfulResponse(resp)) &&
+            (!stream || !stream.didCallback))
```
Comment (AllanFly120) [inline lib/event_listeners.js]: There's no issue here but a little hard to follow.  Will using another if statement be more readable? like this:

```javascript
if (!stream || !stream.didCallback) {
//don't concat response chunks when using event streams unless response is unsuccessful
    if ((AWS.HttpClient.streamsApiVersion === 2) && operation.hasEventOutput && service.successfulResponse(resp) ) {
         return
    }
   resp.request.emit('httpDone');
    done();
}
```
I think it's more readable(?) It's your call.
